Client/Frontend/UserContent/UserScripts/WalletEthereumProvider.js

```diff
@@ -0,0 +1,89 @@
+// Copyright 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+(function() {
+
+function post(method, payload) {
+  return new Promise((resolve, reject) => {
+    webkit.messageHandlers.$<handler>.postMessage({
+      "securitytoken": "$<security_token>",
+      "method": method,
+      "args": JSON.stringify(payload)
+    })
+    .then(resolve, (errorJSON) => {
+      try {
+        reject(JSON.parse(errorJSON))
+      } catch(e) {
+        reject(errorJSON)
+      }
+    })
+  })
+}
+
+Object.defineProperty(window, 'ethereum', {
+  configurable: true,
```

Comment:
If we can't reuse the brave-core definition of this property can we match this object to what's being defined there in order to keep things as closely aligned as possible and make this a bit more maintainable. For example the `writable` and `enumerable` parameters are being set to true rather than the default of false. Ultimately we should align with the goal of making it not possible for a web page to overwrite it like we did with the [solana provider PR](https://github.com/brave/brave-core/pull/12554#discussion_r833722481).

Given this is making RPC calls but brave-core isn't I'm not sure if it's possible to do this.

Reply:
One of the main issues is that the brave-core ethereum provider itself overwrites `window.ethereum` (to be a Proxy) and adds additional properties to it.  Seen here: https://github.com/brave/brave-core/blob/8532627de764fdb25045255eb856c9c71f8e75c8/components/brave_wallet/resources/ethereum_provider.js#L39

Additionally the that redefinition sets `enumerable`, `configurable` and `writable` to true

Reply:
@diracdeltas what's your take on this one? Looking at the comment you left on that other PR in the past we could leave this as is and assume it's untrusted, but it seems a bit odd to be able to trust this on desktop/android but not iOS.

Reply:
just to make sure i understand, are you referring to this comment? https://github.com/brave/brave-core/pull/12554#discussion_r834805673

to clarify, my concern there was around trusted code in brave-core which called `window.ethereum` or `window.solana` so if a page could redefine it or its properties, that could cause security problems. i think at least for solana there is just one instance of this and it's guarded by another check, as mentioned in https://github.com/brave/brave-core/pull/12554#discussion_r835688146. however there is currently no process/check in place (other than manual code review) which prevents people from adding potentially unsafe code in the future, so it's better if we either make sure these objects cannot be redefined or clearly mark them as untrusted.

Reply:
> just to make sure i understand, are you referring to this comment? brave/brave-core#12554 (comment)

Yeah that was the comment I was intending to reference.

> to clarify, my concern there was around trusted code in brave-core which called window.ethereum or window.solana so if a page could redefine it or its properties, that could cause security problems.

Ahh ok that's making more sense to me now why that's a different case then this one. When I originally raised this question I was thinking about this code incorrectly. Based on what I understand now if someone were to override this they'd essentially just be breaking the link here for the JS code to call down to swift code which ultimately calls the brave core code. (and that calls some more JS code to complete the circle of languages 😆 )

In that case I think the solution to preventing this would be what's defined [here](https://github.com/brave/brave-core/pull/12554#discussion_r834841397), but realistically speaking overwriting this would either cause it to break the communication or allow someone to redirect to a different provider.

Reply:
>  I suppose someone could have some fun if they overwrote the isConnected function to break dApp functionality.

wouldn't they just break it for their own site though? it would be a bad problem if a site could override `isConnected` for other origins.

Reply:
They would only break their own site, correct.

Reply:
> > I suppose someone could have some fun if they overwrote the isConnected function to break dApp functionality.
> 
> wouldn't they just break it for their own site though? it would be a bad problem if a site could override `isConnected` for other origins.

Yeah, that's my understanding as well now is this is site isolated so shouldn't cause any problems in that way. I think this one is a non-issue now.